File: 2015/day-2/i-was-told-there-would-be-no-math.js
'use strict';

import { getFilePath, summonTheFile } from '../../util/util.js';

// 2024-12-10 19:10:55
// https://adventofcode.com/2015/day/2

const surBase = 2;

const inputFile = 'sample.txt';
const filePath = getFilePath(import.meta.url, inputFile);
const fileContents = await summonTheFile(filePath);

function main() {
  const areas = make2DArray(fileContents);
  let totalSqft = 0;
  for (const area of areas) {
    totalSqft += calculateComboAreas(area);
  }
  console.log(totalSqft);
}
main();

function make2DArray(fileContents) {
  return fileContents.split('\n').map((s) => s.split('x').map((v) => +v));
}

function calculateComboAreas(list) {
  let totalSqft = 0;
  let smallArea = Infinity;

  for (let i = 0; i < list.length; i++) {
    for (let j = i + 1; j < list.length; j++) {
      const area = list[i] * list[j];

      if (area < smallArea) {
        smallArea = area;
      }
      totalSqft += area * surBase;
    }
  }
  return totalSqft + smallArea;
}
// ans: 1598415

File: 2015/day-2/i-was-told-there-would-be-no-math.go
package day2_2015

import (
	"advent/code/util"
	"log"
	"math"
	"strings"
)

var baseVal = 2

func Main() int {

	cont, err := util.GetFileContents("/2015/day-2/sample.txt")
	if err != nil {
		log.Fatal(err)
	}
	slice2d := make2DSlice(cont)
	total := 0
	for _, row := range slice2d {
		total += calculateAreas(row)
	}
	return total
}

func make2DSlice(contents []byte) [][]int {
	var slice [][]int

	fileContents := strings.Split(string(contents[:]), "\n")
	for _, str := range fileContents {
		trimmed := strings.TrimSpace(str)
		line := strings.Split(strings.TrimSpace(trimmed), "x")
		if len(line) == 1 {
			continue
		}
		var row []int
		for _, val := range line {
			row = append(row, util.StrToInt(val))
		}
		slice = append(slice, row)
	}
	return slice
}

func calculateAreas(areas []int) int {
	totalSqFt := 0
	smallArea := math.MaxInt

	for i := 0; i < len(areas); i++ {
		for j := i + 1; j < len(areas); j++ {
			t := areas[i] * areas[j]
			if t < smallArea {
				smallArea = t
			}
			totalSqFt += t * baseVal

		}
	}
	return totalSqFt + smallArea
}

// ans:1598415
